File: java/Trees/BinarySearchTree.java
package Trees;

import java.util.*;

import definitions.BTree.BTreeNode;

public class BinarySearchTree<T extends Comparable<T>> extends BinaryTree<T> {
    private BTreeNode<T> root;
    private int size;

    public BinarySearchTree() {
        size = 0;
        root = null;
    }

    public void add(T val) {
        root = add(val, root);

    }

    private BTreeNode<T> add(T val, BTreeNode<T> node) {
        if (node == null) {
            size++;
            return new BTreeNode<T>(val);
        }
        if (node.val.compareTo(val) > 0) {
            size++;
            node.left = add(val, node.left);
        } else if (node.val.compareTo(val) < 0) {
            size++;
            node.right = add(val, node.right);
        }
        return node;
    }

    private boolean hasDeleted = false;

    public void remove(T val) {
        hasDeleted = false;
        root = remove(val, root);
        if (hasDeleted) size--;
    }

    private T getMaxValue(BTreeNode<T> node) {
        var max = node.val;
        while (node != null) {
            if (node.val.compareTo(max) > 0) max = node.val;
            node = node.right;
        }
        return max;
    }

    private BTreeNode<T> remove(T val, BTreeNode<T> node) {
        if (node == null) return null;
        if (node.val.compareTo(val) > 0) {
            node.left = remove(val, node.left);
            return node;
        } else if (node.val.compareTo(val) < 0) {
            node.right = remove(val, node.right);
            return node;
        }
        if (node.left == null && node.right == null) {
            hasDeleted = true;
            return null;
        } else if (node.left != null && node.right == null) {
            hasDeleted = true;
            return node.left;
        } else if (node.left == null && node.right != null) {
            hasDeleted = true;
            return node.right;
        } else {
            var maxLeft = getMaxValue(node.left);
            node.val = maxLeft;
            node.left = remove(maxLeft, node.left);
            hasDeleted = true;
            return node;
        }
    }

    public boolean contains(T val) {
        return contains(val, root);
    }

    private boolean contains(T val, BTreeNode<T> node) {
        if (node == null) return false;
        if (node.val.equals(val)) return true;
        if (node.val.compareTo(val) > 0) return contains(val, node.left);
        else return contains(val, node.right);
    }

    public int size() {
        return size;
    }

    public void clear() {
        size = 0;
        root = null;
    }

    public List<T> inOrder() {
        return super.inOrder(root);
    }

    public List<T> preOrder() {
        return super.preOrder(root);
    }

    public List<T> postOrder() {
        return super.postOrder(root);
    }

    public List<List<T>> levelOrder() {
        return super.levelOrder(root);
    }

    public int height() {
        return super.height(root);
    }

    public List<T> rightView() {
        return super.rightView(root);
    }

    public List<T> leftView() {
        return super.leftView(root);
    }

    public List<T> topView() {
        return super.topView(root);
    }

    public List<T> bottomView() {
        return super.bottomView(root);
    }

    public List<T> levelOrderArray() {
        return super.levelOrderArray(root);
    }

    public List<List<T>> verticalOrder() {
        return super.verticalOrder(root);
    }

}

File: typescript/src/trees/binary-search-tree.ts
import type {CompareFn} from "../util.ts";
import {BinaryTree, BinaryTreeNode} from "./binary-tree.ts";

export default class BinarySearchTree<T> extends BinaryTree<T> {
  private root: BinaryTreeNode<T> | null;
  private elements: number;
  private readonly compareFn: CompareFn<T>;

  public constructor(compare: CompareFn<T>) {
    super();
    this.root = null;
    this.elements = 0;
    this.compareFn = compare;
  }

  public add(val: T) {
    this.root = this.addHelper(val, this.root);
  }

  private addHelper(
    val: T,
    node: BinaryTreeNode<T> | null,
  ): BinaryTreeNode<T> | null {
    if (!node) return new BinaryTreeNode(val);
    if (this.compareFn(node.val, val) > 0) {
      node.left = this.addHelper(val, node.left);
      this.elements++;
    } else if (this.compareFn(node.val, val) < 0) {
      node.left = this.addHelper(val, node.right);
      this.elements++;
    }
    return node;
  }

  private hasDeleted: boolean = false;

  public remove(val: T) {
    this.hasDeleted = false;
    this.root = this.removeHelper(val, this.root);
    if (this.hasDeleted) this.elements--;
  }

  private removeHelper(
    val: T,
    node: BinaryTreeNode<T> | null,
  ): BinaryTreeNode<T> | null {
    if (!node) return null;
    if (this.compareFn(node.val, val) > 0) {
      node.left = this.removeHelper(val, node.left);
      return node;
    } else if (this.compareFn(node.val, val) < 0) {
      node.right = this.removeHelper(val, node.right);
      return node;
    } else {
      if (!node.left && !node.right) {
        this.hasDeleted = true;
        return null;
      } else if (!!node.left && !node.right) {
        this.hasDeleted = true;
        return node.left;
      } else if (!node.left && !!node.right) {
        this.hasDeleted = true;
        return node.right;
      } else {
        node.val = this.getMin(node.right);
        this.hasDeleted = true;
        node.right = this.removeHelper(val, node.right);
        return node;
      }
    }
  }

  private getMin(node: BinaryTreeNode<T> | null): T {
    let min = node!.val;
    while (node != null) {
      if (this.compareFn(node.val, min) < 0) min = node.val;
      node = node.left;
    }
    return min;
  }

  public get size() {
    return this.elements;
  }

  public clear() {
    this.root = null;
    this.elements = 0;
  }

  public has(val: T) {
    return this.hasHelper(val, this.root);
  }

  private hasHelper(val: T, node: BinaryTreeNode<T> | null): boolean {
    if (!node) return false;
    if (this.compareFn(node.val, val) == 0) return true;
    const fromLeft = this.hasHelper(val, node.left);
    if (fromLeft) return fromLeft;
    return this.hasHelper(val, node.right);
  }

  public override inOrder(): T[] {
    return super.inOrder(this.root);
  }

  public override preOrder(): T[] {
    return super.preOrder(this.root);
  }

  public override postOrder(): T[] {
    return super.postOrder(this.root);
  }

  public override levelOrder() {
    return super.levelOrder(this.root);
  }

  public override height() {
    return super.height(this.root);
  }

  public override rightView() {
    return super.rightView(this.root);
  }

  public override leftView() {
    return super.leftView(this.root);
  }

  public override topView() {
    return super.topView(this.root);
  }

  public override bottomView() {
    return super.bottomView(this.root);
  }

  public override levelOrderArray() {
    return super.levelOrderArray(this.root);
  }

  public override verticalOrder() {
    return super.verticalOrder(this.root);
  }

  public includes(val: T) {
    return this.includesHelper(val, this.root);
  }

  private includesHelper(val: T, node: BinaryTreeNode<T> | null): boolean {
    if (node == null) return false;
    if (this.compareFn(val, node.val) == 0) return true;
    if (this.includesHelper(val, node.left)) return true;
    return this.includesHelper(val, node.right);
  }
}
